### ISSUE TITLE
Lazy params to return channels and wait

### ISSUE THREAD

**Issue by Shadowfaxenator** [romsar/gonertia#17]
Now lazy params are implemented so that they range in sequence and wait for previous to resolve. To make page loading faster we need to make params functions run as a goroutine and return a channel to wait for all of them

**Comment by romsar** [comment]
https://x.com/pedroborg_es/status/1842190091914969313
Yup, we should wait until next week to see the beta version.
I think there will be a certain amount of work on Gonertia.

**Comment by romsar** [comment]
https://github.com/inertiajs/inertia-laravel/tree/2.x/src
beta 2.0 just came out

**Comment by romsar** [comment]
As I can see, most of the new features have been implemented on the frontend and we don't really need a lot of changes.
Now it seems to me that we only need to make deferred props and history encryption.

**Comment by romsar** [comment]
created 2.x branch

**Comment by romsar** [comment]
I just made Goinertia compatible with Inertia 2.0.

```
go get github.com/romsar/gonertia@2.x
```

All new features are supported:
- [x] merge props
- [x] defer props
- [x] history encryption

Also I wrote tests for all new features and update README.md!

Please check my pull request https://github.com/romsar/gonertia/pull/20 or branch https://github.com/romsar/gonertia/tree/2.x

A few question remain.
I think that the new release version of the package should be 2.0.0.
1. But should we change module name to github.com/romsar/gonertia/v2? At the moment, it seems to me that we should not, but I am not sure that this is the best practice. Any opinions on this?
2. Should we now merge my pr in to master or wait until inertia 2.0 to become stable?

**Comment by romsar** [comment]
@the-goodies if you want you can check gonertia v2 :)

**Comment by JensvandeWiel** [comment]
> * But should we change module name to github.com/romsar/gonertia/v2? At the moment, it seems to me that we should not, but I am not sure that this is the best practice. Any opinions on this?

Hey man, nice work!

You should wait for inertia v2 to release, and maybe its smart to rename the module. It has it's pros and cons

**Comment by the-goodies** [comment]
That was fast, great work. I just tried a little bit and it works great. At least one issue was fixed for me, related to async, when you issue a request and at the same time reload happens (I receive SSE message from the server indicating to refresh/update current screen), this reload doesn't cancel an ongoing request anymore.

Module version should be v2 to reflect that it's specifically for Inertia 2, but whatever you decide is fine by me.
I see no reason to wait until stable release.

**Comment by romsar** [comment]
Just renamed master to 1.x. After some time, we will make 2.x the main branch and release 2.0.0.

**Comment by Shadowfaxenator** [comment]
I don’t see how Optional props are implemented now?  As I see you just range over a slice of props and call func() in sequence. It’s a waterfall, you need to call func() prop as a goroutine and await for all of them to complete with waitgroup. In this case we start to process all func() props in parallel

**Comment by Shadowfaxenator** [comment]
And another thing that’s not clear to me is how you implement new deferred props? Shouldn’t they be called by inertiajs as a separate request? You somehow register new endpoints for them?

**Comment by romsar** [comment]
> I don’t see how Optional props are implemented now?

Now we have interface `ignoreFirstLoad`[ just like in the laravel adapter](https://github.com/inertiajs/inertia-laravel/blob/2.x/src/IgnoreFirstLoad.php).

If prop implemented this interface - it will be resolved only when needed. It was necessary to move this logic into the interface due to the appearance of DeferProp.

> As I see you just range over a slice of props and call func() in sequence. It’s a waterfall, you need to call func() prop as a goroutine and await for all of them to complete with waitgroup. In this case we start to process all func() props in parallel

It's not a bad idea, but I'm not sure if it's worth doing it yet, because we iterate over the map and update its elements. I'm not sure if it can be done competitively without locking mutex on each element. You are free to make a pull request if you know how to do it.

> And another thing that’s not clear to me is how you implement new deferred props? Shouldn’t they be called by inertiajs as a separate request? You somehow register new endpoints for them?

https://v2.inertiajs.com/deferred-props

Deferred props are just same as Optional props, but for them Inertia.js automatically sends a receipt request after page load. There is no need for any additional actions other than returning keys of these props to the front-end.

**Comment by romsar** [comment]
Just pushed new commits to `2.x` branch with same fixes that in https://github.com/inertiajs/inertia-laravel/pull/677
Now flash provider have two new methods to flash and pull clear history flag.

**Comment by romsar** [comment]
gonertia v2.0.0 released
